File: fabric/src/main/java/juuxel/adorn/platform/fabric/NetworkBridgeImpl.java
package juuxel.adorn.platform.fabric;

import juuxel.adorn.platform.NetworkBridge;
import net.fabricmc.fabric.api.networking.v1.PlayerLookup;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.network.protocol.Packet;
import net.minecraft.server.level.ServerPlayer;

public final class NetworkBridgeImpl implements NetworkBridge {
    public static final NetworkBridgeImpl INSTANCE = new NetworkBridgeImpl();

    @Override
    public void sendToTracking(Entity entity, Packet<? super ClientGamePacketListener> packet) {
        for (var player : PlayerLookup.tracking(entity)) {
            ServerPlayNetworking.getSender(player).sendPacket(packet);
        }
    }

    @Override
    public void sendToClient(Player player, CustomPacketPayload payload) {
        if (player instanceof ServerPlayer serverPlayer) {
            ServerPlayNetworking.send(serverPlayer, payload);
        }
    }
}

File: fabric/src/main/kotlin/juuxel/adorn/platform/fabric/NetworkBridgeImpl.kt
package juuxel.adorn.platform.fabric

import juuxel.adorn.fluid.FluidReference
import juuxel.adorn.lib.AdornNetworking
import juuxel.adorn.platform.NetworkBridge
import net.fabricmc.fabric.api.networking.v1.PlayerLookup
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.network.Packet
import net.minecraft.network.listener.ClientPlayPacketListener
import net.minecraft.util.Identifier

object NetworkBridgeImpl : NetworkBridge {
    override fun sendToTracking(entity: Entity, packet: Packet<*>) {
        for (player in PlayerLookup.tracking(entity)) {
            ServerPlayNetworking.getSender(player).sendPacket(packet)
        }
    }

    override fun createEntitySpawnPacket(entity: Entity): Packet<ClientPlayPacketListener> =
        AdornNetworking.createEntitySpawnPacket(entity)

    override fun sendOpenBookPacket(player: PlayerEntity, bookId: Identifier) =
        AdornNetworking.sendOpenBookPacket(player, bookId)

    override fun sendBrewerFluidSync(player: PlayerEntity, syncId: Int, fluid: FluidReference) {
        AdornNetworking.sendBrewerFluidSync(player, syncId, fluid)
    }
}
